File: CodeTop/121. 买卖股票的最佳时机.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        vector<vector<int> > dp(n, vector<int>(2));
        for(int i=0; i<n; i++) {
            if(i - 1 == -1) {
                dp[i][0] = 0;
                dp[i][1] = -prices[0];
                continue;
            }
            dp[i][0] = max(dp[i-1][0], dp[i-1][1] + prices[i]);
            dp[i][1] = max(dp[i-1][1],  - prices[i]);
        }
        return dp[n-1][0];
    }
};

File: 刷题/golang/121. 买卖股票的最佳时机.go
func maxProfit(prices []int) (s int) {
	maxs := math.MinInt64
	mins := math.MaxInt64
	for i := 0; i < len(prices); i++ {
		//记一个最小值
		mins = min(mins, prices[i])
		//记一个最大值
		maxs = max(maxs, prices[i]-mins)

	}
	//返回最大差值
	return maxs
}

func max(x, y int) int {
	if x > y {
		return x
	} else {
		return y
	}
}
func min(x, y int) int {
	if x > y {
		return y
	} else {
		return x
	}
}